### ISSUE TITLE
Molecule lengths inferred from BX tags with BWA

### ISSUE THREAD

**Issue by PA-GAGNAIRE** [pdimens/harpy#3]
Hi Pavel, 

I tested the new conda version of Harpy installed with mamba, and it works great!
I just had to install the R package "circlize" myself in my Harpy environment because apparently it didn't come with the conda installation. 

I'm now working with the BWA mapping pipeline instead of EMA following your suggestion. The alignment works well, as do the various sumstats output by Harpy align, with the exception of the reportGencov.Rmd script, which I had to modify in two places to match the format of the sample.cov.gz file produced with the BWA option (I replaced 'Barcodes' with 'alignments' and modified the circos.genomicTrack() function).

I was wondering if you were planning to integrate in the BWA pipeline the molecule lengths inferred from BX tags (and also possibly binned as a histogram) as you did with the EMA option. We are currently looking into how to extract this information directly from the sorted bam, I'll let you know if we find a solution on our side. 

Best 
Pierre-Alexandre

**Comment by pdimens** [comment]
Thanks for the info! The conda installation isn't stable, I should have mentioned 😅. It was an effort to reserve the name Harpy within bioconda, just in case.
Thanks for pointing out the circlize dependency error and the report error. Those are quick fixes that I should be able to amend today.
There is a python script in `utilities`  (bxStats.py) that is not used by harpy and I'm not yet comfortable with its approach for calculating molecule size, but it's worth you looking at as a jumping off point. I also haven't tried running the perl script on the BWA reads, which probably works. These are great suggestions, and I'm eager to implement them today! I'll keep you posted. Btw, the `dev` branch will now reflect changes after the release.

**Comment by pdimens** [comment]
Currently testing the implementation on the dev branch. Will update in a few hrs

**Comment by pdimens** [comment]
@PA-GAGNAIRE the `dev` branch has this implemented now. The dev branch also including a breaking CLI change where now instead of `--leviathan` or `--bwa` you use the `--method xxxxx` flag. The docstring will provide that info. Also, BWA is now the default mapper on that branch. When things get solidified a little bit more, I will merge it into `main`, create a release, and update the bioconda recipe. If you would like to try the updates:

```bash
# activate the conda env with harpy in it
conda activate <env with harpy>

# clone the harpy repo
git clone https://github.com/pdimens/harpy.git

# switch to dev branch
cd harpy && git checkout dev

## install the new bits
bash misc/buildlocal.sh
```

**Comment by pdimens** [comment]
@PA-GAGNAIRE it took a little bit to code it correctly, but the `dev` branch now has `utilities/bxStats.py` that is my pythonic implementation of the perl script in `utilities/`. They probably don't do the _same exact_ things bc I can't read perl, but the script is annotated kinda-decently and has been giving me very accurate alignment/bx information. The usage is:
```bash
bxStats.py input.bam -c cutoff_value > output.bx.stats
```
The `-c` is the basepair distance at which you would consider alignments with the same BX to not be from the same molecule anymore. This value should reflect your data, but `100000` (default) or `50000` is probably ok. Possibly higher if you've got evidence to support having really long tagged molecules. I would recommend gzipping the output, as it's 25mb compressed for a single sample of my 15x fly data. 

For some context, it will output every BX barcode for each chromosome and provide:
1. chromosome/contig
2. bx ID
3. number of alignments with that bx tag for that chromosome/contig
4. start position of the first alignment for that bx for that chromosome/contig
5. end position of the last alignment for that bx for that chromosome
6. inferred length of the molecule based on the start/end of the alignments with that BX
7. number of aligned basepairs in those alignments for that bx for that chromosome
8. the minimum basepair distance between two alignments sharing a bx (excluding read pairs)

You will notice that there will be barcodes that are suffixed with an integer, e.g. `A11C45B21D96.1`. Those are a way of representing molecules that were split at your `-c` cutoff distance. So using this example, `A11C45B21D96.1` are alignments that have the `A11C45B21D96` barcode but are > `c` base pairs from the last alignment with `A11C45B21D96` (for that contig). The suffixes are sequential, so `A11C45B21D96.2` would be alignments that have the `A11C45B21D96` barcode for that contig but are > `c` base pairs from the last alignment with `A11C45B21D96.1`, which was already marked as split from the original `A11C45B21D96`.  For this to work correctly, your input BAM needs to be position-sorted with `samtools sort`, which harpy already does. 

I'm still working on using the output to make nice informative reports with tables and plots, so `bxStats.py` hasn't yet been integrated into harpy proper. If everything checks out after some more testing and the report code is to my liking, I'll very likely replace the perl script for both ema and bwa mapping methods with this implementation.

**Comment by PA-GAGNAIRE** [comment]
Hi Pavel,

I'm back to Harpy after a 2-weeks break from bioinfo work. I'll try to get back to you when I've made enough progress with the analyses. That's great that you have made a pythonic implementation of synthetic molecule length measure!

Best 
Pierre-Alexandre

**Comment by PA-GAGNAIRE** [comment]
Hi Pavel, 

I've finally been able to test your new dev branch. The new features and the BX Molecule Report are fantastic, thank you for these very useful additions! 

These are minor issues but in order to run the harpy align snakemake pipeline correctly, I had to disable the last commands in trim.smk and align-bwa.smk writing a log file with the "with open()" command, which returned an unknown file error. 

Finally, in the markdown script reportBwaGencov.Rmd, I had to replace expression filter(gencov, contig %in% .contigs)
line 118 with filter(gencov, contig %in% .contigs$contig)

Apart from that, everything runs smoothly and the outputs are very useful for describing the quality of haplotagging for each sample.

I haven't had time to go beyond read mapping yet, but I plan to use the other harpy tools soon. 

Best
Pierre-Alexandre
